internal/app/machined/pkg/controllers/network/internal/probe/probe.go

```diff
@@ -0,0 +1,159 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package probe contains implementation of the network probe runners.
+package probe
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"syscall"
+
+	"github.com/benbjohnson/clock"
+	"github.com/siderolabs/gen/channel"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/pkg/machinery/resources/network"
+)
+
+// Runner describes a state of running probe.
+type Runner struct {
+	ID    string
+	Spec  network.ProbeSpecSpec
+	Clock clock.Clock
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// Notification of a runner status.
+type Notification struct {
+	ID     string
+	Status network.ProbeStatusSpec
+}
+
+// Start a runner with a given context.
+func (runner *Runner) Start(ctx context.Context, notifyCh chan<- Notification, logger *zap.Logger) {
+	runner.wg.Add(1)
+
+	ctx, runner.cancel = context.WithCancel(ctx)
+
+	go func() {
+		defer runner.wg.Done()
+
+		runner.run(ctx, notifyCh, logger)
+	}()
+}
+
+// Stop a runner.
+func (runner *Runner) Stop() {
+	runner.cancel()
+
+	runner.wg.Wait()
+}
+
+// run a probe.
+//
+//nolint:gocyclo
+func (runner *Runner) run(ctx context.Context, notifyCh chan<- Notification, logger *zap.Logger) {
+	logger = logger.With(zap.String("probe", runner.ID))
+
+	if runner.Clock == nil {
+		runner.Clock = clock.New()
+	}
+
+	ticker := runner.Clock.Ticker(runner.Spec.Interval)
+	defer ticker.Stop()
+
+	consecutiveFailures := 0
+	firstIteration := true
+
+	for {
+		if !firstIteration {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+		} else {
+			firstIteration = false
+		}
+
+		err := runner.probe(ctx)
+		if err == nil {
+			if consecutiveFailures > 0 {
+				logger.Info("probe succeeded")
+			}
```

Comment:
didn't get this - we log "probe succeeded" if there is a failure?

Reply:
we log it if there were failures before, so that we don't log each success